text/3857-cfg-version.md

```diff
@@ -0,0 +1,625 @@
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## `since` cfg predicate
+
+We could offer a `before` predicate but that is already covered by `not(since)`.
+
+The `since` name was taken from 
+[rustversion](https://crates.io/crates/rustversion) and the `#[deprecated(since)]` / `#[stable(since)]` attributes.
+This better conveys what operation is being performed than the original `version` name
+and leaves room for related predicates like `before`.
+We could also call this `minimum`, or support comparison operators in the spirit of [RFC 3796](https://github.com/rust-lang/rfcs/pull/3796).
+The risk with a general word like `since` is if we gain support for other data types in cfgs, like integers for embedded development.
+The name `since` might apply in some situations but not others and its unclear if we'd want to generalize it past versions.
+While having a specific name avoids these concerns.
+
+We could swap the order of parameters and make `rust` a default for the second parameter to allow `#[cfg(since("1.95"))]` as a shorthand.
+However, this would look confusing in Cargo and doesn't seem like its offering enough of a benefit to be worth the costs.
+
+The `ConfigurationVersion` is sloppy with the string literal's syntax (relying on `--check-cfg`) so that
+- Allows evolution without requiring an MSRV bump
+- Its consistent with other predicates, e.g. `#[cfg(foo = "1.0")]`
+
+If we were stricter on the syntax,
+we could allow for version numbers to be directly accepted, without quotes 
+(e.g. `#[cfg(since(rust, 1.95.0))]`).
+If we ever decided to support operators (e.g.`#[cfg(since(rust, "=1.95.0"))]`, see `--check-cfg`), then we'd need to decide if those also go outside the string or then require a string, being inconsistent.
+This may limit us if we decide to allow for alternative version formats like with [target_version](#cfg_target_version) as they may not have formats that map well to SemVer.
+Worst case, we'd need to accept arbitrary bare words.
+This would also be inconsistent with other uses of `cfg`s
+*but* maybe that would just be the start to natively supporting more types in `cfg`,
+like integers which are of interest to embedded folks.
+
+## `--check-cfg`
+
+The `--check-cfg` predicate and the value for `rust` ensures users get warnings about
+- Invalid syntax
+- Using this with versions from before its supported, e.g. `#[cfg(since(rust, "1.0.0")]`
+
+`--check-cfg` requires a SemVer version, rather than a version requirement,
+in case we want the future possibility of relaxing SemVer versions
+*and* we want to infer from the fields used in `--check-cfg` to specify the maximum number of fields accepted in comparisons.
+
+We could have the `check-cfg` `since` predicate only apply to the `cfg` `since` predicate,
+causing `#[cfg(rust = "1.100.0")]` to warn.
```

Comment:
I would much prefer that. If someone want's exact matching of version we should IMO have a `#[cfg()]` predicate to do that.

Reply:
Since we talked, do you still feel this way?

I'm going to be writing up a thing for editions, and I was considering it would look like `--check-cfg "cfg(edition, values("2015", "2018", "2021", "2024", since("2025")))"` to say "the versions are discrete but we don't know the future".  For both `edition` and `rust`, we have reasons to do both exact match and version matches